lib/logstash/inputs/tcp.rb

```diff
@@ -86,48 +83,121 @@ def register
         @ssl_context.cert_store = @cert_store
         @ssl_context.verify_mode = OpenSSL::SSL::VERIFY_PEER|OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
       end
-    end # @ssl_enable
+    end
+
+    # note that since we are opening a socket in register, we must also make sure we close it
+    # in the close method even if we also close it in the stop method since we could have
+    # a situation where register is called but not run & stop.
 
     if server?
       @logger.info("Starting tcp input listener", :address => "#{@host}:#{@port}")
       begin
-        @server_socket = TCPServer.new(@host, @port)
+        set_server_socket(TCPServer.new(@host, @port))
       rescue Errno::EADDRINUSE
         @logger.error("Could not start TCP server: Address in use", :host => @host, :port => @port)
         raise
       end
-      if @ssl_enable
-        @server_socket = OpenSSL::SSL::SSLServer.new(@server_socket, @ssl_context)
-      end # @ssl_enable
+
+      set_server_socket(OpenSSL::SSL::SSLServer.new(server_socket, @ssl_context)) if @ssl_enable
+    end
+  end
+
+  def run(output_queue)
+    if server?
+      run_server(output_queue)
+    else
+      run_client(output_queue)
     end
-  end # def register
+  end
+
+  def stop
+    # force close all sockets which will escape any blocking read with a IO exception
+    # and any thread using them will exit
+    server_socket.close rescue nil
```

Comment:
@colinsurprenant The way I read it, your intent is to say "Close unless it's already closed, and then I don't really care about the error" - in this case, I'm ok with `foo.bar rescue nil` to just avoid some mess of exception catching that we're going to ignore anyway. +1 on using 'rescue nil' here since it makes (to me) it clear the intent. Maybe a comment? "Close, but if close fails (already closed?) we don't care" or something